src/Symfony/Component/HttpClient/Internal/AmpListener.php

```diff
@@ -0,0 +1,161 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpClient\Internal;
+
+use Amp\Http\Client\Connection\Stream;
+use Amp\Http\Client\EventListener;
+use Amp\Http\Client\Request;
+use Amp\Promise;
+use Amp\Success;
+use Symfony\Component\HttpClient\Exception\TransportException;
+
+/**
+ * @author Nicolas Grekas <p@tchwork.com>
+ *
+ * @internal
+ */
+class AmpListener implements EventListener
+{
+    private $info;
+    private $pinSha256;
+    private $onProgress;
+
+    public function __construct(array &$info, array $pinSha256, \Closure $onProgress)
+    {
+        $info += [
+            'connect_time' => 0.0,
+            'pretransfer_time' => 0.0,
+            'starttransfer_time' => 0.0,
+            'total_time' => 0.0,
+            'namelookup_time' => 0.0,
+            'primary_ip' => '',
+            'primary_port' => 0,
+        ];
+
+        $this->info = &$info;
+        $this->pinSha256 = $pinSha256;
+        $this->onProgress = $onProgress;
+    }
+
+    public function startRequest(Request $request): Promise
+    {
+        $this->info['start_time'] = $this->info['start_time'] ?? microtime(true);
+        ($this->onProgress)();
+
+        return new Success();
+    }
+
+    public function startDnsResolution(Request $request): Promise
+    {
+        ($this->onProgress)();
+
+        return new Success();
+    }
+
+    public function startConnectionCreation(Request $request): Promise
+    {
+        ($this->onProgress)();
+
+        return new Success();
+    }
+
+    public function startTlsNegotiation(Request $request): Promise
+    {
+        ($this->onProgress)();
+
+        return new Success();
+    }
+
+    public function startSendingRequest(Request $request, Stream $stream): Promise
+    {
+        $host = $stream->getRemoteAddress()->getHost();
+
+        if (false !== strpos($host, ':')) {
+            $host = '['.$host.']';
+        }
+
+        $this->info['primary_ip'] = $host;
+        $this->info['primary_port'] = $stream->getRemoteAddress()->getPort();
+        $this->info['pretransfer_time'] = microtime(true) - $this->info['start_time'];
+
+        if (null !== $tlsInfo = $stream->getTlsInfo()) {
+            foreach ($tlsInfo->getPeerCertificates() as $cert) {
+                $this->info['peer_certificate_chain'][] = openssl_x509_read($cert->toPem());
+            }
```

Comment:
Or `Certificate::getResource()`?

Reply:
Should `$this->info['peer_certificate_chain'][]` be always recorded or just if requested via an option?

Reply:
it *should* be here if the option is set, then it *might* be set of the option is not set :)
so we can make this conditional.

Reply:
That sounds good, because keeping it always enabled is quite some performance bottleneck. My suggestion is to move the pin calculation to `Certificate`, so we don't have to expose the resource. We could add `TlsInfo::getPeerCertificate()` to parse only the leaf certificate. This will keep the `$this->info['peer_certificate_chain'][]` collection a bit less efficient than it could be with exposing the resource, but I guess that's a fair tradeoff.

Reply:
Works for me. Would you work on adding the pinning logic to the class? You can copy/paste from here of course!

Reply:
I can do those, sure.